cpp/ql/src/experimental/Security/CWE/CWE-783/OperatorPrecedenceLogicErrorWhenUseBitwiseOrLogicalOperations.ql

```diff
@@ -0,0 +1,179 @@
+      instanceof BoolType and
+    not exptmp
+        .(LogicalAndExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "0" and
+    not exptmp
+        .(LogicalAndExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "1"
+  )
+}
+
+/** Holds if `exptmp` equals expression bit operations in reverse priority order. */
+predicate isBitwiseandBitwise(Expr exptmp) {
+  not exptmp.(BitwiseOrExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(BitwiseOrExpr).getRightOperand().isParenthesised() and
+  (
+    exptmp.(BitwiseOrExpr).getRightOperand() instanceof BitwiseAndExpr or
+    exptmp.(BitwiseOrExpr).getRightOperand() instanceof BitwiseXorExpr
+  )
+  or
+  not exptmp.(BitwiseXorExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(BitwiseXorExpr).getRightOperand().isParenthesised() and
+  exptmp.(BitwiseXorExpr).getRightOperand() instanceof BitwiseAndExpr
+}
+
+/** Holds if the range contains no boundary values. */
+predicate isRealRange(Expr exp) {
+  upperBound(exp).toString() != "18446744073709551616" and
+  upperBound(exp).toString() != "9223372036854775807" and
+  upperBound(exp).toString() != "4294967295" and
+  upperBound(exp).toString() != "Infinity" and
+  upperBound(exp).toString() != "NaN" and
+  lowerBound(exp).toString() != "-9223372036854775808" and
+  lowerBound(exp).toString() != "-4294967296" and
+  lowerBound(exp).toString() != "-Infinity" and
+  lowerBound(exp).toString() != "NaN" and
+  upperBound(exp) != 2147483647 and
+  upperBound(exp) != 268435455 and
+  upperBound(exp) != 33554431 and
+  upperBound(exp) != 8388607 and
+  upperBound(exp) != 65535 and
+  upperBound(exp) != 32767 and
+  upperBound(exp) != 255 and
+  upperBound(exp) != 127 and
+  lowerBound(exp) != -2147483648 and
+  lowerBound(exp) != -268435456 and
+  lowerBound(exp) != -33554432 and
+  lowerBound(exp) != -8388608 and
+  lowerBound(exp) != -65536 and
+  lowerBound(exp) != -32768 and
+  lowerBound(exp) != -128
+  or
+  lowerBound(exp) = 0 and
+  upperBound(exp) = 1
+}
+
+/** Holds if expressions are of different size or range */
+predicate isDifferentSize(Expr exp1, Expr exp2, Expr exp3) {
+  exp1.getType().getSize() = exp2.getType().getSize() and
+  exp1.getType().getSize() != exp3.getType().getSize()
+  or
+  (
+    isRealRange(exp1) and
+    isRealRange(exp2) and
+    isRealRange(exp3)
+  ) and
+  upperBound(exp1).maximum(upperBound(exp2)) - upperBound(exp1).minimum(upperBound(exp2)) < 16 and
+  lowerBound(exp1).maximum(lowerBound(exp2)) - lowerBound(exp1).minimum(lowerBound(exp2)) < 16 and
+  (
+    upperBound(exp1).maximum(upperBound(exp3)) - upperBound(exp1).minimum(upperBound(exp3)) > 256 or
+    lowerBound(exp1).maximum(lowerBound(exp2)) - lowerBound(exp1).minimum(lowerBound(exp2)) > 256
+  )
+}
+
+/** Holds if it is possible to get different values of the expression */
```

Comment:
Which of the 5 expression parameters is this comment referring to?

Reply:
Thank you.
now I see that the comment is not very correct.
perhaps this one is more informative.
`Holds if different values of the expression obtained from the parameters of the predicate can be obtained.`

Reply:
That sounds more precise, indeed.